### PR TITLE
Fix analyzer issues

### PR DESCRIPTION
**Pull request by reuvenperetz**
## Pull Request Description:
Issues:
1. The float graph used for comparison in the similarity analyzer was a graph that underwent stats correction.
2. The number of samples to display did not work as expected, and a single sample was always used.
3. Error was thrown when no compare points were found.

This PR fixes these issues by:
1. Changing the analyzer_model_quantization API to receive both float and quantized graphs.
2. Add a method to the visualizer to query the existence of compare points and skip the analyzer with a warning in case no points were found.
3. Change the computation of the number of visualized samples.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/ptq/keras/quantization_facade.py [hunk 1]
```diff
@@ -140,15 +142,20 @@ def keras_post_training_quantization(in_model: Model,
                                                target_resource_utilization=target_resource_utilization,
                                                tb_w=tb_w)
 
-        tg = ptq_runner(tg, representative_data_gen, core_config, fw_info, fw_impl, tb_w)
+        float_graph = copy.deepcopy(tg)
+
+        graph_with_stats_correction = ptq_runner(tg, representative_data_gen, core_config, fw_info, fw_impl, tb_w)
+        quantized_graph = quantize_graph_weights(graph_with_stats_correction)
```
Comment (ofirgo) [inline model_compression_toolkit/ptq/keras/quantization_facade.py:148]: Can this run only if we use similarity analyzer? (inside the if scope beneath that)

File: model_compression_toolkit/core/common/quantization/quantize_graph_weights.py [hunk 2]
```diff
@@ -23,20 +23,21 @@
 from model_compression_toolkit.logger import Logger
 
 
-def quantize_graph_weights(graph: Graph) -> Graph:
+def quantize_graph_weights(graph_to_quantize: Graph) -> Graph:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/quantize_graph_weights.py:26]: maybe this is for a separate PR, but I don't like this function, especially since it is under common/quantization.
It can be mistakenly considered as THE function for quantizing the graph, where in practice it is only used for the analyzer (and for second-moment correction, which I think is by mistake and also is used nowhere).

To sum up - I think we should move this method to be under the analyzer package, give it proper documentation for its use, and fix the second moment so it won't use it.
If you agree then consider if you want to do this in this PR or in a separate one.

File: model_compression_toolkit/ptq/pytorch/quantization_facade.py [hunk 3]
```diff
@@ -112,16 +115,20 @@ def pytorch_post_training_quantization(in_module: Module,
                                                target_resource_utilization=target_resource_utilization,
                                                tb_w=tb_w)
 
-        tg = ptq_runner(tg, representative_data_gen, core_config, DEFAULT_PYTORCH_INFO, fw_impl, tb_w)
+        float_graph = copy.deepcopy(tg)
+
+        graph_with_stats_correction = ptq_runner(tg, representative_data_gen, core_config, DEFAULT_PYTORCH_INFO, fw_impl, tb_w)
+        quantized_graph = quantize_graph_weights(graph_with_stats_correction)
```
Comment (ofirgo) [inline model_compression_toolkit/ptq/pytorch/quantization_facade.py:121]: same (under the if)

File: model_compression_toolkit/ptq/pytorch/quantization_facade.py [hunk 4]
```diff
@@ -112,16 +115,20 @@ def pytorch_post_training_quantization(in_module: Module,
                                                target_resource_utilization=target_resource_utilization,
                                                tb_w=tb_w)
 
-        tg = ptq_runner(tg, representative_data_gen, core_config, DEFAULT_PYTORCH_INFO, fw_impl, tb_w)
+        float_graph = copy.deepcopy(tg)
+
+        graph_with_stats_correction = ptq_runner(tg, representative_data_gen, core_config, DEFAULT_PYTORCH_INFO, fw_impl, tb_w)
```
Comment (ofirgo) [inline model_compression_toolkit/ptq/pytorch/quantization_facade.py:120]: Please put DEFAULT_PYTORCH_INFO into a variable and use the variable same way as in the Keras facade?

File: model_compression_toolkit/ptq/keras/quantization_facade.py [hunk 5]
```diff
@@ -140,15 +142,20 @@ def keras_post_training_quantization(in_model: Model,
                                                target_resource_utilization=target_resource_utilization,
                                                tb_w=tb_w)
 
-        tg = ptq_runner(tg, representative_data_gen, core_config, fw_info, fw_impl, tb_w)
+        float_graph = copy.deepcopy(tg)
```
Comment (ofirgo) [inline model_compression_toolkit/ptq/keras/quantization_facade.py:145]: But this is not the float graph, it's the graph that is outputted from the core runner..
What am I missing?

(relevant to all facades so I'm not writing this again)

File: model_compression_toolkit/gptq/pytorch/quantization_facade.py [hunk 6]
```diff
@@ -193,7 +196,12 @@ def pytorch_gradient_post_training_quantization(model: Module,
                                  hessian_info_service=hessian_info_service)
 
         if core_config.debug_config.analyze_similarity:
-            analyzer_model_quantization(representative_data_gen, tb_w, graph_gptq, fw_impl, DEFAULT_PYTORCH_INFO)
+            analyzer_model_quantization(representative_data_gen,
+                                        tb_w,
+                                        float_graph,
+                                        graph_gptq,
+                                        fw_impl,
+                                        DEFAULT_PYTORCH_INFO)
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/pytorch/quantization_facade.py:204]: same regarding DEFAULT_PYTORCH_INFO into variable

File: model_compression_toolkit/gptq/keras/quantization_facade.py [hunk 7]
```diff
@@ -225,7 +229,12 @@ def keras_gradient_post_training_quantization(in_model: Model, representative_da
         del hessian_info_service
 
         if core_config.debug_config.analyze_similarity:
-            analyzer_model_quantization(representative_data_gen, tb_w, tg_gptq, fw_impl, fw_info)
+            analyzer_model_quantization(representative_data_gen,
+                                        tb_w,
+                                        float_graph,
+                                        tg_gptq,
+                                        fw_impl,
+                                        DEFAULT_KERAS_INFO)
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/keras/quantization_facade.py:237]: same (variable)

File: model_compression_toolkit/core/common/visualization/nn_visualizer.py [hunk 8]
```diff
@@ -57,24 +56,38 @@ class NNVisualizer:
...
-        self.graph_quantized = quantize_graph_weights(graph_float)
+        self.graph_quantized = graph_quantized
         self.fw_impl = fw_impl
         self.fw_info = fw_info
 
         # Get compare points of two graphs.
         self.compare_points, self.compare_points_name = _get_compare_points(self.graph_quantized)
         self.compare_points_float, self.compare_points_name_float = _get_compare_points(self.graph_float)
 
+        assert len(self.compare_points) == len(
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/visualization/nn_visualizer.py:81]: why not logger?

File: model_compression_toolkit/core/analyzer.py [hunk 9]
```diff
@@ -41,23 +42,32 @@ def analyzer_model_quantization(representative_data_gen: Callable,
...
-
-        for i, _data in enumerate(representative_data_gen()):
-            if i >= NUM_SAMPLES_DISTANCE_TENSORBOARD:
-                break
-            figure = visual.plot_distance_graph(_data,
-                                                distance_fn=compute_cs,
-                                                convert_to_range=lambda a: 1 - 2 * a)
-            tb_w.add_figure(figure, f'similarity_distance_sample_{i}')
+        if not visual.has_compare_points():
+            Logger.warning(f'No comparing points were found to plot analyze similarity.')
```
Comment (ofirgo) [inline model_compression_toolkit/core/analyzer.py:57]: Change to logger.error (informing an error but continuing running, I think it's more suitable here_

File: model_compression_toolkit/core/analyzer.py [hunk 10]
```diff
@@ -41,23 +42,32 @@ def analyzer_model_quantization(representative_data_gen: Callable,
...
+                    if visualized_samples >= NUM_SAMPLES_DISTANCE_TENSORBOARD:
+                        break
+                    figure = visual.plot_distance_graph(_data,
+                                                        sample_index=sample_index,
+                                                        distance_fn=compute_cs,
+                                                        convert_to_range=lambda a: 1 - 2 * a)
+                    tb_w.add_figure(figure, f'similarity_distance_sample_{visualized_samples}')
+                    visualized_samples += 1
+            if visualized_samples < NUM_SAMPLES_DISTANCE_TENSORBOARD:
+                Logger.warning(f'Not enough batches in representative dataset to generate {NUM_SAMPLES_DISTANCE_TENSORBOARD} figures')
```
Comment (ofirgo) [inline model_compression_toolkit/core/analyzer.py:72]: Same, use error (it would continue to run but would inform the user for the occurred error)

**Review by ofirgo — CHANGES_REQUESTED** [review]

**Review by elad-c — APPROVED** [review]

File: model_compression_toolkit/core/common/quantization/quantize_graph_weights.py (hunk 2 above)
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/quantization/quantize_graph_weights.py:26]: Since second-moment also depends on it, we should think about it in a separate PR.

**Review by ofirgo — APPROVED** [review]
